### PR TITLE
SAK-48745 LTI Implement Course Groups Service

### PR DESCRIPTION
**Pull request by csev**

### CODE REVIEW

File: basiclti/basiclti-blis/src/java/org/sakaiproject/lti13/LTI13Servlet.java [hunk 1]
```diff
@@ -1577,102 +1591,105 @@ protected void handleGroupService(String site_id, HttpServletRequest request, Ht
...
 
-            response.setContentType(LTI13ConstantsUtil.CONTEXTGROUPCONTAINER_TYPE);
-            out = response.getWriter();
-            out.println("{");
-            out.println(" \"id\" : \""+currentUrl+"\", ");
-            out.println(" \"groups\": [");
-
-            for (Iterator i = groups.iterator();i.hasNext();) {
-                Group group = (Group) i.next();
+			for (Iterator i = groups.iterator();i.hasNext();) {
```
Comment (ottenhoff) [inline basiclti/basiclti-blis/src/java/org/sakaiproject/lti13/LTI13Servlet.java:1660]:
```suggestion
			for (Group group : groups) {
```

File: basiclti/basiclti-blis/src/java/org/sakaiproject/lti13/LTI13Servlet.java [hunk 2]
```diff
@@ -929,14 +937,25 @@ protected void handleTokenPost(String tool_id, HttpServletRequest request, HttpS
...
 			}
 			returnScopeSet.add(LTI13ConstantsUtil.SCOPE_NAMES_AND_ROLES);
 
 			sat.addScope(SakaiAccessToken.SCOPE_ROSTER);
 		}
 
+		if (scope.contains(LTI13ConstantsUtil.SCOPE_CONTEXTGROUP_READONLY)) {
+			if (allowRoster != 1) {
+				LTI13Util.return400(response, "invalid_scope", LTI13ConstantsUtil.SCOPE_CONTEXTGROUP_READONLY);
+				log.error("Scope context grup not allowed {}", tool_id);
```
Comment (zhuoY121) [inline basiclti/basiclti-blis/src/java/org/sakaiproject/lti13/LTI13Servlet.java:951]: Typo: "grup"

File: basiclti/basiclti-blis/src/java/org/sakaiproject/lti13/LTI13Servlet.java [hunk 3]
```diff
@@ -1488,18 +1506,26 @@ protected void handleNamesAndRoles(String signed_placement, HttpServletRequest r
 				*/
 				sakai_ext.put("sakai_role", sakaiRole);
 
+				// TODO: Should we only do this if requested via groups=true ??
```
Comment (zhuoY121) [inline basiclti/basiclti-blis/src/java/org/sakaiproject/lti13/LTI13Servlet.java:1509]: Do we need the "TODO" comment?

File: basiclti/tsugi-util/src/java/org/tsugi/lti13/objects/GroupService.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+       "service_versions": ["1.0"]
+   }
+
+*/
+public class GroupService extends org.tsugi.jackson.objects.JacksonBase {
+    public static String SCOPE_CONTEXTGROUP_READONLY = "https://purl.imsglobal.org/spec/lti-gs/scope/contextgroup.readonly";
+
+    @JsonProperty("scope")
+    public List<String> scope = new ArrayList<String>();
+	@JsonProperty("context_groups_url")
```
Comment (zhuoY121) [inline basiclti/tsugi-util/src/java/org/tsugi/lti13/objects/GroupService.java:31]: Spacing

File: basiclti/tsugi-util/src/test/org/tsugi/lti13/GroupServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.core.JsonProcessingException;
+
+/**
+ *
+ * @author csev
+ */
+public class GroupServiceTest {
+
+	@Test
```
Comment (zhuoY121) [inline basiclti/tsugi-util/src/test/org/tsugi/lti13/GroupServiceTest.java:25]: Spacing

**Review by zhuoY121 — APPROVED** [review]

File: basiclti/basiclti-blis/src/java/org/sakaiproject/lti13/LTI13Servlet.java (hunk 3 above)
Comment (csev) [reply basiclti/basiclti-blis/src/java/org/sakaiproject/lti13/LTI13Servlet.java:1509]: Yes - The sakai extension always included the groups - now that there is an official way to include the groups, I wonder if we should delete the sakai_extension.  MY theory is that I will keep them as if for a bit, let the new serivce find its place, then deprecate the sakai_extension and then follow the spec 100% (will improve performance).  But for now I want to rember to come back and do this.

**Pull request merged**